### PR TITLE
[Cards] Add Cards dynamic color support

### LINKED ISSUE #7878: [Cards] Internal issue: b/137184241
This was filed as an internal issue. If you are a Googler, please visit [b/137184241](http://b/137184241) for more details.

---
#### Internal data
- Associated internal bug: [b/137184241](http://b/137184241)

### PR DESCRIPTION
**Pull request by andrewoverton**
This PR simply makes it so that MDCCard calls updateShadowColor in layoutSubviews and adds a snapshot test.
Closes #7878.

### CODE REVIEW

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m [hunk 1]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
```
Comment (romoore) [inline components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m]: Nit: Please fix the copyright date.

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MaterialSnapshot.h"
+
+#import "MaterialCards+Theming.h"
+#import "MaterialCards.h"
+
+@interface MDCCardWithCustomTraitCollection : MDCCard
```
Comment (romoore) [inline components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m:23]: Nit: Please add docs for the interface.

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+@property(nonatomic, strong) UITraitCollection *traitCollectionOverride;
+@end
+
+@implementation MDCCardWithCustomTraitCollection
+- (UITraitCollection *)traitCollection {
+  return self.traitCollectionOverride ?: [super traitCollection];
+}
+@end
+
+@interface MDCCardCustomTraitCollectionSnapshotTests : MDCSnapshotTestCase
```
Comment (romoore) [inline components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m:37]: Nit: Please add docs for the interface.

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+- (void)generateSnapshotAndVerifyForView:(UIView *)view {
+  UIView *snapshotView = [view mdc_addToBackgroundView];
+  [self snapshotVerifyView:snapshotView];
+}
+
+- (void)testShadowColorRespondsToDynamicColor {
+#if defined(__IPHONE_13_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_13_0)
+  if (@available(iOS 13.0, *)) {
+    // Given
+    UIColor *darkModeColor = UIColor.whiteColor;
```
Comment (romoore) [inline components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m]: Nit: Please don't create a local variable - you can just use `UIColor.whiteColor` directly in the `else` clause.

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    UIColor *dynamicColor =
+        [UIColor colorWithDynamicProvider:^(UITraitCollection *traitCollection) {
+          if (traitCollection.userInterfaceStyle == UIUserInterfaceStyleLight) {
+            return UIColor.blackColor;
+          } else {
+            return darkModeColor;
```
Comment (romoore) [inline components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m]: Please make the shadow color something very obviously different from the normal/default value. I would suggest something like `UIColor.magentaColor`.  I would also recommend increasing the elevation of the card so the shadow is easier to see in the image.

**Review by romoore — APPROVED** [review]

File: components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m (hunk 1 above)
Comment (andrewoverton) [reply components/Cards/tests/snapshot/MDCCardCustomTraitCollectionSnapshotTests.m]: Omg I keep doing it! >_<

**Review by yarneo — APPROVED** [review]

**Pull request merged**